# Interactive Business Dashboard in Streamlit

## Introduction

Business dashboards provide an interactive way to analyze sales performance and support data-driven decision-making. Instead of examining raw data, dashboards summarize key business metrics through interactive charts, filters, and Key Performance Indicators (KPIs).

In this project, an interactive business dashboard is developed using Streamlit and the Global Superstore dataset. The dashboard allows users to explore sales, profit, customer performance, and product categories through dynamic filters and visualizations. It enables business users to quickly identify trends, compare regional performance, and make informed decisions based on real-time insights.

## Problem Statement

Businesses generate large volumes of sales data every day. Analyzing this information using spreadsheets alone is time-consuming and makes it difficult to identify important trends and business opportunities.

To improve decision-making, an interactive dashboard is required that enables users to filter sales data by Region, Category, and Sub-Category while displaying important business metrics such as Total Sales, Total Profit, Top Customers, and sales performance through interactive visualizations.

## Project Objectives

The main objectives of this project are:

- Perform data cleaning and preprocessing.
- Prepare the dataset for dashboard development.
- Develop an interactive dashboard using Streamlit.
- Implement interactive filters for Region, Category, and Sub-Category.
- Display key performance indicators (KPIs) such as Total Sales, Total Profit, Total Orders, and Total Customers.
- Visualize business performance using interactive charts.
- Identify the Top 5 Customers by Sales.
- Support business decision-making through interactive data storytelling.

## Dataset Information

This project uses the **Global Superstore Dataset**, which is widely used for business intelligence, sales analytics, and dashboard development projects.

**Dataset Source:**  
https://www.kaggle.com/datasets/apoorvaappz/global-super-store-dataset

The dataset contains detailed sales transaction records from a global retail superstore, including customer information, product details, order history, shipping data, and financial metrics.

The primary objective of this dataset is to analyze business performance, identify sales trends, evaluate regional and product-wise performance, and support data-driven business decision-making.

Each record represents an individual sales transaction and includes information such as customer details, product category, region, market, sales, profit, quantity, discount, shipping cost, and order priority.

This dataset is well suited for developing **interactive business intelligence dashboards** using **Streamlit**, enabling users to explore sales performance through dynamic filters, key performance indicators (KPIs), and interactive visualizations.

### Dataset Summary

| Attribute | Description |
|------------|-------------|
| Dataset Name | Global Superstore Dataset |
| Source | Kaggle |
| File Used | Global_Superstore.csv |
| Problem Type | Business Intelligence Dashboard |
| Dashboard Framework | Streamlit |
| Domain | Retail and Sales |

### Dataset Features

| Feature | Description |
|----------|-------------|
| Row ID | Unique identifier for each sales record |
| Order ID | Unique identifier for each customer order |
| Order Date | Date on which the order was placed |
| Ship Date | Date on which the order was shipped |
| Ship Mode | Shipping method used for delivery |
| Customer ID | Unique identifier for each customer |
| Customer Name | Name of the customer |
| Segment | Customer segment (Consumer, Corporate, or Home Office) |
| City | City where the customer is located |
| State | State or province where the customer is located |
| Country | Country where the order was placed |
| Postal Code | Postal or ZIP code of the customer's location |
| Market | Market in which the sale occurred |
| Region | Geographic region of the order |
| Product ID | Unique identifier for each product |
| Category | Main product category |
| Sub-Category | Product sub-category |
| Product Name | Name of the product |
| Sales | Total sales amount generated from the order |
| Quantity | Number of units sold |
| Discount | Discount applied to the order |
| Profit | Profit earned from the order |
| Shipping Cost | Shipping cost for the order |
| Order Priority | Priority level assigned to the order |

## Import Required Libraries

Before starting the data preparation process, the required Python libraries are imported.

These libraries are used for data manipulation, data cleaning, date handling, and exploratory analysis before developing the interactive Streamlit dashboard.

- **Pandas** is used for data manipulation and analysis.
- **NumPy** is used for numerical computations.
- **Matplotlib** and **Seaborn** are used for exploratory data visualization during data preparation.

In [55]:
# Data Manipulation

import pandas as pd
import numpy as np

# Data Visualization

import matplotlib.pyplot as plt
import seaborn as sns

import warnings

warnings.filterwarnings("ignore")

## Load Dataset

The Global Superstore dataset is loaded into a Pandas DataFrame for analysis and data preparation.

A copy of the original dataset is also created to preserve the raw data while performing data cleaning and preprocessing. The cleaned dataset will later be used to build the interactive Streamlit dashboard.

In [56]:
# Load the dataset

df = pd.read_csv(
    "../dataset/Global_Superstore.csv",
    encoding="latin1"
)

# Create a copy of the original dataset

df_clean = df.copy()

## Dataset Preview

After loading the dataset, the first few records are displayed to verify that the dataset has been loaded successfully and to gain an initial understanding of its structure.

In [57]:
# Display the first five records

df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,City,State,...,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit,Shipping Cost,Order Priority
0,32298,CA-2012-124891,31-07-2012,31-07-2012,Same Day,RH-19495,Rick Hansen,Consumer,New York City,New York,...,TEC-AC-10003033,Technology,Accessories,Plantronics CS510 - Over-the-Head monaural Wir...,2309.650,7,0.0,762.1845,933.57,Critical
1,26341,IN-2013-77878,05-02-2013,07-02-2013,Second Class,JR-16210,Justin Ritter,Corporate,Wollongong,New South Wales,...,FUR-CH-10003950,Furniture,Chairs,"Novimex Executive Leather Armchair, Black",3709.395,9,0.1,-288.7650,923.63,Critical
2,25330,IN-2013-71249,17-10-2013,18-10-2013,First Class,CR-12730,Craig Reiter,Consumer,Brisbane,Queensland,...,TEC-PH-10004664,Technology,Phones,"Nokia Smart Phone, with Caller ID",5175.171,9,0.1,919.9710,915.49,Medium
3,13524,ES-2013-1579342,28-01-2013,30-01-2013,First Class,KM-16375,Katherine Murray,Home Office,Berlin,Berlin,...,TEC-PH-10004583,Technology,Phones,"Motorola Smart Phone, Cordless",2892.510,5,0.1,-96.5400,910.16,Medium
4,47221,SG-2013-4320,05-11-2013,06-11-2013,Same Day,RH-9495,Rick Hansen,Consumer,Dakar,Dakar,...,TEC-SHA-10000501,Technology,Copiers,"Sharp Wireless Fax, High-Speed",2832.960,8,0.0,311.5200,903.04,Critical


## Data Cleaning and Preparation

Data cleaning and preparation are essential steps before developing an interactive business dashboard.

In this phase, the dataset is examined for missing values, duplicate records, data types, and data consistency to ensure accurate analysis and reliable dashboard visualizations.

### Dataset Shape

The shape of the dataset is examined to determine the total number of records and features available for business analysis and dashboard development.

In [58]:
# Display dataset shape

df.shape

(51290, 24)

#### Observation

The dataset contains **51,290 records** and **24 features**, providing comprehensive sales, customer, product, and regional information for business analysis and interactive dashboard development.

### Dataset Information

The dataset information is examined to understand the structure of the dataset, including column names, data types, and non-null values. This helps identify potential data quality issues before data preparation.

In [59]:
# Display dataset information

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 51290 entries, 0 to 51289
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Row ID          51290 non-null  int64  
 1   Order ID        51290 non-null  str    
 2   Order Date      51290 non-null  str    
 3   Ship Date       51290 non-null  str    
 4   Ship Mode       51290 non-null  str    
 5   Customer ID     51290 non-null  str    
 6   Customer Name   51290 non-null  str    
 7   Segment         51290 non-null  str    
 8   City            51290 non-null  str    
 9   State           51290 non-null  str    
 10  Country         51290 non-null  str    
 11  Postal Code     9994 non-null   float64
 12  Market          51290 non-null  str    
 13  Region          51290 non-null  str    
 14  Product ID      51290 non-null  str    
 15  Category        51290 non-null  str    
 16  Sub-Category    51290 non-null  str    
 17  Product Name    51290 non-null  str    
 1

#### Observation

The dataset contains **24 features** with a combination of categorical, numerical, and date-related variables. Most columns have **51,290 non-null values**, indicating complete data. However, the **Postal Code** column contains missing values and will require further examination during the data cleaning process. The dataset includes **17 categorical features, 5 floating-point features, and 2 integer features**, making it suitable for business analysis and dashboard development.

### Missing Values

The dataset is examined for missing values to identify incomplete records that may affect business analysis and dashboard accuracy. Detecting missing values helps determine whether they should be removed, replaced, or retained based on their significance.

In [60]:
# Check missing values

df.isnull().sum()

Row ID                0
Order ID              0
Order Date            0
Ship Date             0
Ship Mode             0
Customer ID           0
Customer Name         0
Segment               0
City                  0
State                 0
Country               0
Postal Code       41296
Market                0
Region                0
Product ID            0
Category              0
Sub-Category          0
Product Name          0
Sales                 0
Quantity              0
Discount              0
Profit                0
Shipping Cost         0
Order Priority        0
dtype: int64

#### Observation

The missing values analysis shows that **only the Postal Code column contains missing values (41,296)**, while all other **23 features are complete**. Since the dashboard focuses on sales, profit, customer, region, category, and sub-category analysis, the missing postal codes are not expected to significantly affect the business insights.

#### Handling Missing Values

The Postal Code column contains a large number of missing values (41,296), representing approximately 80% of the dataset.

Since the Postal Code feature is not required for the dashboard filters, KPI calculations, or business analysis, no imputation or removal is performed. Retaining the original values preserves the integrity of the dataset and avoids introducing inaccurate information.

### Duplicate Records

The dataset is examined for duplicate records to ensure that each sales transaction is unique. Removing duplicate entries helps maintain data accuracy and prevents misleading business insights in the interactive dashboard.

In [61]:
# Check duplicate records

print("Duplicate Records: ",df.duplicated().sum())

Duplicate Records:  0


#### Observation

No duplicate records were found in the dataset. Therefore, no duplicate rows need to be removed before preparing the data for dashboard development.

### Data Consistency Check

The dataset is examined for common data quality issues such as leading and trailing spaces in text-based features. Ensuring consistent text values improves filtering, grouping, and visualization accuracy within the interactive dashboard.

#### Check for Leading and Trailing Spaces

The dataset is examined for leading and trailing spaces across all text-based features. Removing unnecessary spaces ensures consistent categorical values and improves the accuracy of filtering, grouping, and dashboard visualizations.

In [62]:
# Check for leading and trailing spaces in all text columns

text_columns = df.select_dtypes(include=["object"]).columns

space_summary = pd.DataFrame(
    [
        [column, (df[column] != df[column].str.strip()).sum()]
        for column in text_columns
    ],
    columns=["Column Name", "Rows with Extra Spaces"]
)

space_summary[space_summary["Rows with Extra Spaces"] > 0]

,Column Name,Rows with Extra Spaces
15,Product Name,16


##### Observation

The data consistency check identified leading and trailing spaces only in the **Product Name** column, affecting **16 records**. All other text-based features were free from unnecessary spaces.

#### Remove Leading and Trailing Spaces

Leading and trailing spaces are removed from all text-based features to ensure consistent categorical values. This improves filtering, grouping, and visualization accuracy within the interactive dashboard.

In [63]:
# Remove leading and trailing spaces from all text columns

text_columns = df_clean.select_dtypes(include=["object"]).columns

for column in text_columns:
    df_clean[column] = df_clean[column].str.strip()

In [64]:
# Verify that no leading and trailing spaces remain

space_summary = pd.DataFrame(
    [
        [column, (df_clean[column] != df_clean[column].str.strip()).sum()]
        for column in text_columns
    ],
    columns=["Column Name", "Rows with Extra Spaces"]
)

space_summary[space_summary["Rows with Extra Spaces"] > 0]

,Column Name,Rows with Extra Spaces


##### Observation

Leading and trailing spaces were successfully removed from all text-based features. No columns contain unnecessary spaces, ensuring standardized values for accurate filtering, grouping, and dashboard analysis.

### Date Format Conversion

The **Order Date** and **Ship Date** columns are converted from text format to the datetime format. This conversion enables time-based analysis such as yearly trends, monthly sales, quarterly reports, and date filtering within the interactive Streamlit dashboard.

In [65]:
# Convert date columns to datetime format

df_clean["Order Date"] = pd.to_datetime(
    df_clean["Order Date"],
    dayfirst=True
)

df_clean["Ship Date"] = pd.to_datetime(
    df_clean["Ship Date"],
    dayfirst=True
)

In [66]:
# Verify updated data types

df_clean[["Order Date", "Ship Date"]].dtypes

Order Date    datetime64[us]
Ship Date     datetime64[us]
dtype: object

#### Observation

The **Order Date** and **Ship Date** columns were successfully converted from string format to the **datetime** data type. This enables efficient time-based analysis, including yearly and monthly sales trends, and supports interactive date filtering in the Streamlit dashboard.

### Feature Engineering

Feature engineering is performed to create additional date-related features from the **Order Date** column.

These features simplify time-based analysis and improve the performance of the interactive Streamlit dashboard by enabling efficient filtering and visualization of yearly, quarterly, monthly, and daily sales trends.

In [67]:
# Create date-related features

df_clean["Order Year"] = df_clean["Order Date"].dt.year

df_clean["Order Quarter"] = df_clean["Order Date"].dt.quarter

df_clean["Order Month"] = df_clean["Order Date"].dt.month

df_clean["Order Month Name"] = df_clean["Order Date"].dt.month_name()

df_clean["Order Day"] = df_clean["Order Date"].dt.day

df_clean["Order Day Name"] = df_clean["Order Date"].dt.day_name()

In [68]:
# Display the newly created features

df_clean[
    [
        "Order Date",
        "Order Year",
        "Order Quarter",
        "Order Month",
        "Order Month Name",
        "Order Day",
        "Order Day Name"
    ]
].head()

,Order Date,Order Year,Order Quarter,Order Month,Order Month Name,Order Day,Order Day Name
0,2012-07-31,2012,3,7,July,31,Tuesday
1,2013-02-05,2013,1,2,February,5,Tuesday
2,2013-10-17,2013,4,10,October,17,Thursday
3,2013-01-28,2013,1,1,January,28,Monday
4,2013-11-05,2013,4,11,November,5,Tuesday


#### Observation

Six additional date-related features were successfully created from the **Order Date** column. These features enable efficient yearly, quarterly, monthly, and daily analysis, improving the flexibility and interactivity of the Streamlit business dashboard.

### Final Dataset Verification

After completing all data cleaning and feature engineering steps, the final dataset is verified to ensure that it is complete, consistent, and ready for exploratory data analysis and dashboard development.

In [69]:
# Display dataset information

df_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 51290 entries, 0 to 51289
Data columns (total 30 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Row ID            51290 non-null  int64         
 1   Order ID          51290 non-null  str           
 2   Order Date        51290 non-null  datetime64[us]
 3   Ship Date         51290 non-null  datetime64[us]
 4   Ship Mode         51290 non-null  str           
 5   Customer ID       51290 non-null  str           
 6   Customer Name     51290 non-null  str           
 7   Segment           51290 non-null  str           
 8   City              51290 non-null  str           
 9   State             51290 non-null  str           
 10  Country           51290 non-null  str           
 11  Postal Code       9994 non-null   float64       
 12  Market            51290 non-null  str           
 13  Region            51290 non-null  str           
 14  Product ID        51290 non-null 

##### Observation

The final dataset verification confirms that all preprocessing and feature engineering steps were successfully completed. The dataset now contains **51,290 records** and **30 features**, including six newly created date-related features. The **Order Date** and **Ship Date** columns were successfully converted to the datetime format, making the dataset suitable for time-based analysis and interactive dashboard development. The only remaining missing values are in the **Postal Code** column, which were intentionally retained as they do not affect the business analysis or dashboard requirements.

#### Statistical Summary

Summary statistics are generated for the key numerical business features to examine their distribution and verify that the cleaned dataset contains valid values before exploratory data analysis and dashboard development.

In [72]:
# Display summary statistics for numerical business features

df_clean[
    ["Sales", "Quantity", "Discount", "Profit", "Shipping Cost"]
].describe()

,Sales,Quantity,Discount,Profit,Shipping Cost
count,51290.000000,51290.000000,51290.000000,51290.000000,51290.000000
mean,246.490581,3.476545,0.142908,28.610982,26.375915
std,487.565361,2.278766,0.212280,174.340972,57.296804
min,0.444000,1.000000,0.000000,-6599.978000,0.000000
25%,30.758625,2.000000,0.000000,0.000000,2.610000
50%,85.053000,3.000000,0.000000,9.240000,7.790000
75%,251.053200,5.000000,0.200000,36.810000,24.450000
max,22638.480000,14.000000,0.850000,8399.976000,933.570000


##### Observation

The statistical summary confirms that the key business metrics contain valid and reasonable values. The **Sales**, **Quantity**, **Discount**, **Profit**, and **Shipping Cost** columns exhibit realistic distributions suitable for business analysis. Negative values in the **Profit** column represent financial losses and are considered valid business outcomes. Overall, no additional preprocessing is required before proceeding to exploratory data analysis and dashboard development.

#### Date Range Verification

The minimum and maximum values of the date columns are examined to verify that the date conversion was successful and to determine the time period covered by the dataset.

In [73]:
# Check date ranges

print("Order Date Range:")
print(df_clean["Order Date"].min())
print(df_clean["Order Date"].max())

print("\nShip Date Range:")
print(df_clean["Ship Date"].min())
print(df_clean["Ship Date"].max())

Order Date Range:
2011-01-01 00:00:00
2014-12-31 00:00:00

Ship Date Range:
2011-01-03 00:00:00
2015-01-07 00:00:00


##### Observation

The date range verification confirms that the **Order Date** values span from **January 1, 2011** to **December 31, 2014**, while the **Ship Date** values range from **January 3, 2011** to **January 7, 2015**. The later ship dates are expected because orders placed at the end of 2014 were delivered in early 2015. This confirms that the date conversion was successful and the dataset is suitable for time-based analysis and dashboard development.

#### Verify Important Categorical Features

The important categorical features are examined to verify their unique values, identify the available categories, and ensure that no missing values remain after the data cleaning process. This verification confirms that the categorical variables are complete and suitable for filtering and business analysis within the interactive dashboard.

In [83]:
# Summary of important categorical features

pd.set_option("display.max_colwidth", None)

categorical_columns = [
    "Ship Mode",
    "Segment",
    "Market",
    "Region",
    "Country",
    "State",
    "Category",
    "Sub-Category",
    "Order Priority"
]

categorical_summary = pd.DataFrame({
    "Feature": categorical_columns,
    "Unique Values": [df_clean[col].nunique() for col in categorical_columns],
    "Categories": [
        ", ".join(sorted(df_clean[col].dropna().astype(str).unique()))
        if df_clean[col].nunique() <= 20
        else f"Not displayed ({df_clean[col].nunique()} categories)"
        for col in categorical_columns
    ],
    "Missing Values": [df_clean[col].isnull().sum() for col in categorical_columns]
})

categorical_summary

,Feature,Unique Values,Categories,Missing Values
0,Ship Mode,4,"First Class, Same Day, Second Class, Standard Class",0
1,Segment,3,"Consumer, Corporate, Home Office",0
2,Market,7,"APAC, Africa, Canada, EMEA, EU, LATAM, US",0
3,Region,13,"Africa, Canada, Caribbean, Central, Central Asia, EMEA, East, North, North Asia, Oceania, South, Southeast Asia, West",0
4,Country,147,Not displayed (147 categories),0
5,State,1094,Not displayed (1094 categories),0
6,Category,3,"Furniture, Office Supplies, Technology",0
7,Sub-Category,17,"Accessories, Appliances, Art, Binders, Bookcases, Chairs, Copiers, Envelopes, Fasteners, Furnishings, Labels, Machines, Paper, Phones, Storage, Supplies, Tables",0
8,Order Priority,4,"Critical, High, Low, Medium",0


##### Observation

The verification confirms that all important categorical features contain valid and complete values with no missing records. Features with a limited number of categories are displayed explicitly, while high-cardinality features such as **Country** and **State** are summarized to maintain readability. The verified categorical features are suitable for filtering, grouping, and interactive business analysis within the Streamlit dashboard.

### Save Cleaned Dataset

The cleaned dataset is saved as a new CSV file after completing all preprocessing and feature engineering steps. Saving the processed dataset ensures that the cleaning process does not need to be repeated and provides a consistent data source for exploratory data analysis and the Streamlit dashboard.

In [84]:
# Save the cleaned dataset

df_clean.to_csv(
    "../dataset/Global_Superstore_Cleaned.csv",
    index=False,
    encoding="utf-8-sig"
)

print("Cleaned dataset saved successfully.")

Cleaned dataset saved successfully.
